_posts/2018-05-29-Introducing-Bazel-Cquery.md

```diff
@@ -0,0 +1,90 @@
+| Performance          | faster, less acurate                                                       | slower, accurate                     |                                                                                 |
+| Functions            | siblings, buildfiles, tests                                                | config                               | allpaths, attrs, dep, filter, kind, labels, loadfiles, rdeps, somepath, visible |
+| Output Formats       | build, label,  label_kind, minrank, maxrank, location, package, graph, xml | label_and_configuration, transitions | proto                                                                           |
+| Options              | query options                                                              | cquery options, build options        | common query options                                                            |
+
+So, if your priority is speed and over-approximation of results isn’t a problem, `query` is your engine. If your priority is results that match a specific bazel invocation’s flags and fancy output formats aren’t too important to you, then `cquery` is the better choice.
+
+\* Bazel also supports the ‘Sky Query’ engine which automatically kicks in with a specific set of options used with ‘query’. It supports a few extra functions and in some circumstances may be faster and less memory-intensive than ‘query’
+
+## Some Motivating Examples
+
+The library //third_party/zlib:zlibonly isn't in the BUILD file for //foo:foo, but it is an indirect dependency. Why?
+
+```
+$ bazel query "somepath(//foo:foo, third_party/zlib:zlibonly)"
+//foo:foo
+//translations/tools:translator
+//translations/base:base
+//third_party/py/MySQL:MySQL
+//third_party/py/MySQL:_MySQL.so
+//third_party/mysql:mysql
+//third_party/zlib:zlibonly
+```
+
+The [Bazel query how-to](https://docs.bazel.build/versions/master/query-how-to.html) contains many common examples of how `query` is used. Many of the examples can also apply to `cquery`. In the following example, we see where `cquery`’s strength lies (properly resolving a [select](https://docs.bazel.build/versions/master/be/functions.html#select) statement).
+
+```
+$ cat > tree/BUILD <<EOF
+sh_library(
+    name = "ash",
+    deps = select({
+        ":excelsior": [":manna-ash"],
+        ":americana": [":white-ash"],
+        "//conditions:default": [":common-ash"],
+    }),
+)
+sh_library(name = "manna-ash")
+sh_library(name = "white-ash")
+sh_library(name = "common-ash")
+config_setting(
+    name = "excelsior",
+    values = {"define": "species=excelsior"},
+)
+config_setting(
+    name = "americana",
+    values = {"define": "species=americana"},
+)
+
+# Traditional query
+$ bazel query "deps(//tree:ash)" --define species=excelsior
+...error because query does not understand --define…
```

Comment:
Be aware that your editor "helpfully" replaced the trailing ... here with an ellipsis character (but not the first one), which is probably not what you want.